File: lib/test/structuredfield/parseDictionary.test.ts
import assert from 'node:assert';
import test from 'node:test';
import { SfItem } from '../../src/structuredfield/SfItem.js';
import { parseDict } from '../../src/structuredfield/parse/parseDict.js';

test('parseDictionary', () => {
	assert.deepStrictEqual(parseDict(
		`int=1, dec=1.23, token=a, str="a", bool=?1, bin=:AQID:, date=@1659578233`,
	), {
		value: {
			'int': new SfItem(1),
			'dec': new SfItem(1.23),
			'token': new SfItem(Symbol.for('a')),
			'str': new SfItem('a'),
			'bool': new SfItem(true),
			'bin': new SfItem(new Uint8Array([1, 2, 3])),
			'date': new SfItem(new Date(1659578233 * 1000)),
		},
		src: ``,
	});

	assert.deepEqual(parseDict(`a=?0, b, c; foo=bar`), {
		value: {
			'a': new SfItem(false),
			'b': new SfItem(true),
			'c': new SfItem(true, { 'foo': Symbol.for('bar') }),
		},
		src: ``,
	});

	assert.deepStrictEqual(parseDict(`rating=1.5, feelings=(joy sadness)`), {
		value: {
			'rating': new SfItem(1.5),
			'feelings': new SfItem([Symbol.for('joy'), Symbol.for('sadness')]),
		},
		src: ``,
	});

	assert.deepStrictEqual(parseDict(`a=(1 2), b=3, c=4;aa=bb, d=(5 6);valid`), {
		value: {
			'a': new SfItem([1, 2]),
			'b': new SfItem(3),
			'c': new SfItem(4, { 'aa': Symbol.for('bb') }),
			'd': new SfItem([5, 6], { 'valid': true }),
		},
		src: ``,
	});

	assert.throws(() => parseDict(`a=1&`), /failed to parse "&" as Dict/);
	assert.throws(() => parseDict(`a=1,`), /failed to parse "" as Dict/);
});

File: python/src/test/structuredfield/parseDictionary_test.py
import unittest
from datetime import datetime

from ...structuredfield.parse.parseDict import parseDict
from ...structuredfield.parse.ParsedValue import ParsedValue
from ...structuredfield.SfException import SfException
from ...structuredfield.SfItem import SfItem
from ...structuredfield.SfToken import SfToken


class TestDecodeSfItem(unittest.TestCase):
    def test_parseDictionary(self):
        self.assertEqual(
            parseDict(
                'int=1, dec=1.23, token=a, str="a", bool=?1, bin=:AQID:, date=@1659578233'
            ),
            ParsedValue(
                {
                    "int": SfItem(1),
                    "dec": SfItem(1.23),
                    "token": SfItem(SfToken("a")),
                    "str": SfItem("a"),
                    "bool": SfItem(True),
                    "bin": SfItem(b"\x01\x02\x03"),
                    "date": SfItem(datetime.fromtimestamp(1659578233)),
                },
                "",
            ),
        )

        self.assertEqual(
            parseDict("a=?0, b, c; foo=bar"),
            ParsedValue(
                {
                    "a": SfItem(False),
                    "b": SfItem(True),
                    "c": SfItem(True, {"foo": SfToken("bar")}),
                },
                "",
            ),
        )

        self.assertEqual(
            parseDict("rating=1.5, feelings=(joy sadness)"),
            ParsedValue(
                {
                    "rating": SfItem(1.5),
                    "feelings": SfItem([SfToken("joy"), SfToken("sadness")]),
                },
                "",
            ),
        )

        self.assertEqual(
            parseDict("a=(1 2), b=3, c=4;aa=bb, d=(5 6);valid"),
            ParsedValue(
                {
                    "a": SfItem([1, 2]),
                    "b": SfItem(3),
                    "c": SfItem(4, {"aa": SfToken("bb")}),
                    "d": SfItem([5, 6], {"valid": True}),
                },
                "",
            ),
        )

        self.assertRaises(SfException, parseDict, "a=1&")
        self.assertRaises(SfException, parseDict, "a=1,")
